### PR TITLE
List surveys in system console

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Added API to list surveys to be displayed in system console.

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-58123

### CODE REVIEW

File: webapp/src/components/systemConsole/surveyResults/surveyResults.tsx [hunk 1]
```diff
@@ -12,74 +12,45 @@ import type {SurveyResult} from 'types/plugin';
...
+            surveyResult.endDate = new Date(surveyResult.startTime);
+            surveyResult.endDate.setDate(surveyResult.endDate.getDate() + surveyResult.duration);
+
+            surveyResult.npsScore = calculateNPS(surveyResult.promoterCount, surveyResult.passiveCount, surveyResult.detractorCount);
         });
-        setSurveyResults(results);
     }, []);
 
+    useEffect(() => {
+        const work = async () => {
```
Comment (M-ZubairAhmed) [inline webapp/src/components/systemConsole/surveyResults/surveyResults.tsx:38]: can we be more descriptive on what `work` means?

File: webapp/src/components/systemConsole/surveyResults/surveyResults.tsx [hunk 2]
```diff
@@ -12,74 +12,45 @@ import type {SurveyResult} from 'types/plugin';
...
-        ];
+    const [surveyResults, setSurveyResults] = useState<SurveyResult[]>([]);
+
+    const calculateNPS = (promoters: number, passives: number, detractors: number): number => {
+        const totalResponses = promoters + passives + detractors;
+
+        const promoterPercentage = (promoters / totalResponses) * 100;
+        const detractorPercentage = (detractors / totalResponses) * 100;
+
+        return Math.round(promoterPercentage - detractorPercentage);
```
Comment (M-ZubairAhmed) [inline webapp/src/components/systemConsole/surveyResults/surveyResults.tsx:23]: does this have to be inside of component? It can be outside as well right

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/src/components/systemConsole/surveyResults/surveyResults.tsx (hunk 1 above)
Comment (harshilsharma63) [reply webapp/src/components/systemConsole/surveyResults/surveyResults.tsx]: Done. Renamed.

File: webapp/src/components/systemConsole/surveyResults/surveyResults.tsx (hunk 2 above)
Comment (harshilsharma63) [reply webapp/src/components/systemConsole/surveyResults/surveyResults.tsx]: Yeah, no need for this to be inside the component. Moved it to outside the component now.

**Review by M-ZubairAhmed — APPROVED** [review]

File: server/app/survey_response.go [hunk 3]
```diff
@@ -160,3 +231,109 @@ func (a *UserSurveyApp) UpdatePostForExpiredSurvey(userID, surveyID string) erro
...
+		oldPromoterFactor, oldNeutralFactor, oldDetractorFactor = a.getRatingGroupFactors(oldRating)
+
+		oldPromoterFactor *= -1
+		oldNeutralFactor *= -1
+		oldDetractorFactor *= -1
+	}
+
+	var promoterFactor, neutralFactor, detractorFactor int
+
+	if oldPromoterFactor != 0 {
```
Comment (devinbinnie) [inline server/app/survey_response.go:286]: Nit: could separate this into a function to avoid repetition

**Review by devinbinnie — APPROVED** [review]

File: server/app/survey_response.go (hunk 3 above)
Comment (harshilsharma63) [reply server/app/survey_response.go]: Done. Added a `CoalesceInt` function to get first non 0 number from a list of number. The new code look like this-

```go
promoterFactor := utils.CoalesceInt(oldPromoterFactor, newPromoterFactor)
neutralFactor := utils.CoalesceInt(oldNeutralFactor, newNeutralFactor)
detractorFactor := utils.CoalesceInt(oldDetractorFactor, newDetractorFactor)
```

File: server/store/survey_stat.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	return surveyStats, nil
+}
+
+func (s *SQLStore) surveyStatColumns() []string {
+	surveyStateColumns := []string{
+		"receipt_count",
+		"response_count",
+		"passives_count",
+		"promoters_count",
+		"detractors_count",
```
Comment (esarafianou) [inline server/store/survey_stat.go:41]: @harshilsharma63 what are the promoters and the detractors?

File: server/app/survey_response.go [hunk 5]
```diff
@@ -131,6 +161,49 @@ func (a *UserSurveyApp) sendAcknowledgementPost(userID, surveyPostID string) err
...
+		}
+
+		for responseQuestionID := range response.Response {
+			if _, ok := surveyQuestionIDMap[responseQuestionID]; !ok {
+				return errors.New("invalid question ID found in submitted answer")
+			}
+		}
+
+		response.ResponseType = model.ResponseTypeComplete
+	}
```
Comment (esarafianou) [inline server/app/survey_response.go:202]: @harshilsharma63 this defers from the previous implementation in that the linear response no longer seems required in the response length is > 1. 

Isn't that the opposite of what's desired? (always check that the linear question is answered)

Comment (esarafianou) [reply server/app/survey_response.go:202]: This is what happens if you send a response with complete responseType and without the linear question

Screenshot 2024-05-20 at 15 08 50

Comment (esarafianou) [reply server/app/survey_response.go:202]: The issue is that it's marked as a valid response, in the survey stats.

**Pull request merged**